src/core/execute.c

```diff
@@ -2526,17 +2507,33 @@ static int setup_keyring(
 
                         return log_unit_error_errno(u, r, "Failed to link user keyring into session keyring: %m");
                 }
+        }
 
-                if (uid_is_valid(uid) && uid != saved_uid) {
-                        if (setreuid(saved_uid, -1) < 0) {
-                                (void) setregid(saved_gid, -1);
-                                return log_unit_error_errno(u, errno, "Failed to change UID back for user keyring: %m");
-                        }
+        /* Restore uid/gid back */
+        if (uid_is_valid(uid) && uid != saved_uid) {
+                if (setreuid(saved_uid, -1) < 0) {
+                        (void) setregid(saved_gid, -1);
+                        return log_unit_error_errno(u, errno, "Failed to change UID back for user keyring: %m");
```

Comment:
I am just realizing a buglet here which existed in the old code and should be fixed at the same time i figure: the errno might be clobbered by the setregid() here, but we complain about setreuid() here. Hence, to be fully correct, please rearrange this to be:

```c
if (setreuid(saved_uid, -1)  < 0) {
        r = -errno;
        (void) setregid(saved_gid, -1);
        return log_unit_error_errno(u, r, …);
}
```